.github/workflows/benchmark.yml

```diff
@@ -0,0 +1,175 @@
+# This workflow is the main workflow for regenerating the benchmarks data needed for Bowtie's UI.
+# It runs all benchmarks over Bowtie's supported implementations, publishing the benchmark reports for use in the frontend.
+name: Collect New Benchmark Results
+
+on:
+  workflow_dispatch:
+  schedule:
+    # Every Monday at 08:00 UTC
+    - cron: "0 8 * * 1"
+
+permissions:
+  contents: read
+  pages: write
+  id-token: write
+
+jobs:
+  dialects:
+    runs-on: ubuntu-latest
+    outputs:
+      dialects: ${{ steps.dialects-matrix.outputs.dialects }}
+    steps:
+      - uses: actions/checkout@v4
+      - name: Collect supported dialects
+        id: dialects-matrix
+        run: |
+          printf 'dialects=%s\n' "$(jq -c '[.[].shortName]' data/dialects.json)" >> $GITHUB_OUTPUT
```

Comment:
The reason we use this (already inscrutible) code is because in the CI job if we didn't (and used the alternative, `bowtie filter-dialects`) we would have to wait some amount of time to install bowtie and run it before doing useful work (running the suite).

Here, that seems anyways the case -- your next job waits for this one, and then the job after that waits for the middle one -- and the middle one installs bowtie and runs some inscrutible shell code.

Point being, if you anyways are doing some stuff installing bowtie before doing anything useful, you should just do this in one job and using bowtie's CLI.

(Ideally your entire shell script should be even doable using Bowtie's CLI in some trivial way without needing to write 12 lines of stuff, but that's possibly less trivial to fix.)

Reply:
Hmm you are right, here `bowtie filter-dialects` itself can be used.

Reply:
I tried testing with this, it seems like `bowtie filter-dialects` supports only URI format in output. I am using this dialect name to  differentiate different dialect benchmarks as well. The URI contains characters which github doesnt allow to be present in an artifact name.

Currently I am using: `pattern: benchmark-file-${{ matrix.dialect }}-*` to filter all benchmark reports of a particular dialect.

Shall I continue using this or switch to using `bowtie filter-dialects` and then apply some string manipulation to the URI while naming the file to avoid this error ? @Julian

Reply:
(If we do end up leaving as is, leave a comment then that changing this depends on #1332)

So wait do you not use this for anything else other than naming the artifact, which you then ignore and use just to collect the artifact? Or do you really use the short name itself somehow too?

If it's literally just to be able to upload multiple artifacts I think again you should have a look, surely there's a per-matrix identifier that github makes available for every subjob.

But if you don't find one or if you're using it somehow, then yeah leave as is. Thanks for checking.

Reply:
Its basically just helping group artifacts

Say we have:

benchmark-dialect2020-1.json
benchmark-dialect2020-2.json
benchmark-dialect2019-1.json
benchmark-dialect2019-2.json


The shortname is just being used to then filter the benchmark

Reply:
The shortname is also being used to get the benchmarks supported for each dialect but for that the URI works too so there is no issue in that.

Reply:
Also to keep the naming convention of benchmark reports similar to test-reports, the shortname is required.

Reply:
So ig it would be better to do this once #1332 is done.

Reply:
> Also to keep the naming convention of benchmark reports similar to test-reports, the shortname is required.

That doesn't seem the same to me -- test reports are by their nature specific to a dialect, and compliance is always relative to a version of the spec.

A benchmark really isn't, the fact that dialects come into the picture at all is simply to know whether a particular dialect is supported so we know whether we can run a specific schema against an implementation, but I would prefer the notion of a dialect be nowhere near the real "structure" of benchmark reports and just be a kind of 'tag' like any other that we can maybe filter by at some point, but the main structure of benchmarks I think should just be "flat" if that makes sense in an imprecise sense.

Reply:
Ig we had discussed this earlier that it would be great that at sometime we support say running benchmark over the same implementation and same schema but diff dialects and see the difference. So in a way benchmark reports are tied to dialects.

Reply:
Maybe once multi dialect supported within the same benchmark is added then this would be possible

Reply:
OK, I'm good with whatever here to start!

Reply:
Yeah once that is supported , displaying it on the UI would also require some thinking so as to how to show those comparisons. Because now its easy and similar to tests that just show how dialect2020 test of `implementation1`  performs against other implementations on the same dialect. But when multi dialects are supported we would need to think on what would `implementation1` performing better than `implementation2` mean if its not tied to a particular dialect.  Would it mean on avg of all tests of all dialects or what. This would need to be thought of. Idk if I made much sense :(

Reply:
Unless we have a good reason to be more complicated, I1 faster than I2 should mean that the benchmarks run by I1 are a superset of the benchmarks run by I2 (not necessarily a strict superset) and that the geometric average of runtime over that subset of benchmarks is faster in I1 than I2 -- but yeah again in my opinion with no reference to dialects whatsoever, you just flatten all of them and consider them individually.

Reply:
So currently in the frontend we have separate pages for each dialect, I should combine them all into one then ?

Reply:
Ok then ig we can let this be like this for now and once the above is supported we can flatten it all.